src/ring_theory/valuation/valuation_ring.lean

```diff
@@ -0,0 +1,343 @@
+/-
+Copyright (c) 2022 Adam Topaz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Adam Topaz
+-/
+import ring_theory.valuation.integers
+import ring_theory.ideal.local_ring
+import ring_theory.localization.fraction_ring
+import ring_theory.discrete_valuation_ring
+import tactic.field_simp
+
+/-!
+# Valuation Rings
+
+A valuation ring is a domain such that for every pair of elements `a b`, either `a` divides
+`b` or vice-versa.
+
+Any valuation ring induces a natural valuation on its fraction field, as we show in this file.
+Namely, given the following instances:
+`[comm_ring A] [is_domain A] [valuation_ring A] [field K] [algebra A K] [is_fraction_ring A K]`,
+there is a natural valuation `valuation A K` on `K` with values in `value_group A K` where
+the image of `A` under `algebra_map A K` agrees with `(valuation A K).integer`.
+
+We also show that valuation rings are local and that their lattice of ideals is totally ordered.
+-/
+
+universes u v w
+
+/-- An integral domain is called a `valuation ring` provided that for any pair
+of elements `a b : A`, either `a` divides `b` or vice versa. -/
+class valuation_ring (A : Type u) [comm_ring A] [is_domain A] : Prop :=
+(cond [] : ∀ a b : A, ∃ c : A, a * c = b ∨ b * c = a)
```

Comment:
Two questions:
Why not use `a ∣ b ∨ b ∣ a` ?
Would it be better to bundle `is_domain`, i.e. `class valuation_ring (A : Type u) [comm_ring A] extends is_domain A`?

Reply:
Using `a | b \or b | a` means that in practice you need to split that into two cases, and for each one obtain the `c` such that a * c = b` (or vice versa). With the current approch, you first obtain `c` once and for all, then split, saving some lines of code.
One should not really use this condition anyway once the API is set up (e.g. it is much better to use the actual valuation!).

The typeclass assumptions I used mimic those for `discrete_valuation_ring`.